src/lib/storage/segment_access_counter.hpp

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <map>
+#include <memory>
+#include <string>
+
+#include "storage/pos_list.hpp"
+#include "types.hpp"
+
+namespace opossum {
+class Table;
+
+// The SegmentAccessCounter is a collection of counters to register how often and how a segment is accessed.
+class SegmentAccessCounter {
+  friend class SegmentAccessCounterTest;
+
+ public:
+  using CounterType = std::atomic_uint64_t;
+
+  enum class AccessType {
+    Point /* Single point access */,
+    Sequential /* 0, 1, 1, 2, 3, 4 */,
+    Increasing /* 0, 0, 1, 2, 4, 8, 17 */ /* or decreasing */,
+    Random /* 0, 1, 0, 42 */,
+    Dictionary,
+    Count /* Dummy entry to describe the number of elements in this enum class. */
+  };
+
+  inline static const std::array<std::string, (size_t)AccessType::Count> access_type_string_mapping = {
+    "Point", "Sequential", "Increasing", "Random", "Dictionary"
+  };
+
+  SegmentAccessCounter();
+  SegmentAccessCounter(const SegmentAccessCounter& counter);
+  SegmentAccessCounter& operator=(const SegmentAccessCounter& counter);
+
+  CounterType& get(const AccessType type);
+  const CounterType& get(const AccessType type) const;
+
+  static AccessType access_type(const PosList& positions);
+
+  uint64_t sum() const;
```

Comment:
Still not sure what this is good for. In tests, you can also compare against `EXPECT_EQ(..., {1, 5, 4});`

Reply:
The function can be nice if you just want to check if the segment has been accessed at all.

Reply:
`!= ArrayType{}`?

Reply:
That is a pretty cool idea.